File: backend/jobs_store.py
"""
jobs_store.py — Thread-safe in-memory job store.

Every processing job gets a UUID and is stored here.  Background threads
(Demucs, Whisper, Claude) write progress updates to this dict; the
frontend polls GET /api/status/{job_id} to read them.
"""

import threading
from typing import Any, Dict, Optional

# The global job registry — maps job_id → job dict
jobs: Dict[str, Dict[str, Any]] = {}

# A simple re-entrant lock so background threads can safely update jobs
_lock = threading.RLock()


# ─── Step labels ─────────────────────────────────────────────────────────────
STEP_UPLOADING   = "Uploading"
STEP_ISOLATING   = "Isolating Vocals"
STEP_TRANSCRIBING = "Transcribing"
STEP_TRANSLATING = "Translating"
STEP_SYNCING     = "Syncing Lyrics"
STEP_READY       = "Ready"


# ─── CRUD ────────────────────────────────────────────────────────────────────

def create_job(job_id: str, metadata: Optional[Dict] = None) -> Dict:
    """
    Initialise a new job entry with default values.
    Call this immediately after the file is received.
    """
    job = {
        "id":          job_id,
        "status":      "processing",   # processing | complete | error
        "progress":    0,              # 0–100
        "step":        STEP_UPLOADING,
        "error":       None,
        "result":      None,           # final lyric JSON (when complete)
        "file_path":   None,           # local path to uploaded audio
        "vocals_path": None,           # path to isolated vocals
        "segments":    None,           # raw Whisper segments
        "waveform":    None,           # downsampled amplitude array for WaveformPlayer
        "metadata":    metadata or {}, # title, artist, language, etc.
    }
    with _lock:
        jobs[job_id] = job
    return job


def update_job(job_id: str, **kwargs) -> None:
    """
    Update any fields on an existing job.
    Safe to call from background threads.
    """
    with _lock:
        if job_id in jobs:
            jobs[job_id].update(kwargs)


def get_job(job_id: str) -> Optional[Dict]:
    """Return the full job dict, or None if unknown."""
    with _lock:
        job = jobs.get(job_id)
        # Return a shallow copy so callers can't accidentally mutate shared state
        return dict(job) if job else None


def set_error(job_id: str, message: str) -> None:
    """Mark a job as failed with an error message."""
    update_job(job_id, status="error", error=message)


def set_complete(job_id: str, result: Any) -> None:
    """Mark a job as complete with the final lyrics result."""
    update_job(
        job_id,
        status="complete",
        progress=100,
        step=STEP_READY,
        result=result,
    )

File: backend/src/main/java/com/lyrictranslate/JobsStore.java
package com.lyrictranslate;

import com.lyrictranslate.model.Job;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * JobsStore — Thread-safe in-memory job registry.
 *
 * Direct Java equivalent of jobs_store.py.
 * Uses ConcurrentHashMap instead of Python's threading.RLock + dict.
 *
 * Spring singleton bean — injected wherever job state is needed.
 */
@Component
public class JobsStore {

    // Step label constants (mirrors jobs_store.py)
    public static final String STEP_UPLOADING    = "Uploading";
    public static final String STEP_ISOLATING    = "Isolating Vocals";
    public static final String STEP_TRANSCRIBING = "Transcribing";
    public static final String STEP_TRANSLATING  = "Translating";
    public static final String STEP_SYNCING      = "Syncing Lyrics";
    public static final String STEP_READY        = "Ready";

    private final ConcurrentHashMap<String, Job> jobs = new ConcurrentHashMap<>();

    // ── CRUD ────────────────────────────────────────────────────────────────

    /**
     * Initialise a new job entry with default values.
     * Equivalent of jobs_store.create_job().
     */
    public Job createJob(String jobId, Map<String, Object> metadata) {
        Job job = new Job(jobId, metadata);
        jobs.put(jobId, job);
        return job;
    }

    /**
     * Return the Job, or null if unknown.
     * Equivalent of jobs_store.get_job().
     */
    public Job getJob(String jobId) {
        return jobs.get(jobId);
    }

    /**
     * Update progress + step label.
     * Thread-safe: ConcurrentHashMap guarantees visibility across threads.
     */
    public void updateProgress(String jobId, int progress, String step) {
        Job job = jobs.get(jobId);
        if (job != null) {
            job.progress = progress;
            if (step != null) job.step = step;
        }
    }

    public void updateFilePath(String jobId, String filePath) {
        Job job = jobs.get(jobId);
        if (job != null) job.filePath = filePath;
    }

    public void updateVocalsPath(String jobId, String vocalsPath) {
        Job job = jobs.get(jobId);
        if (job != null) job.vocalsPath = vocalsPath;
    }

    public void updateMetadata(String jobId, Map<String, Object> metadata) {
        Job job = jobs.get(jobId);
        if (job != null) job.metadata = metadata;
    }

    public void updateSegments(String jobId, Object segments) {
        Job job = jobs.get(jobId);
        if (job != null) job.segments = segments;
    }

    /**
     * Mark a job as failed.
     * Equivalent of jobs_store.set_error().
     */
    public void setError(String jobId, String message) {
        Job job = jobs.get(jobId);
        if (job != null) {
            job.status = "error";
            job.error  = message;
        }
    }

    /**
     * Mark a job as complete with the final lyrics result.
     * Equivalent of jobs_store.set_complete().
     */
    public void setComplete(String jobId, Object result) {
        Job job = jobs.get(jobId);
        if (job != null) {
            job.status   = "complete";
            job.progress = 100;
            job.step     = STEP_READY;
            job.result   = result;
        }
    }
}
